internal/controller/components/kserve/kserve_controller_actions.go

```diff
@@ -267,33 +266,34 @@ func cleanUpTemplatedResources(ctx context.Context, rr *odhtypes.ReconciliationR
 
 	logger := logf.FromContext(ctx)
 
-	// Removed or unmanaged: remove from rr.Resources
-	if rr.DSCI.Spec.ServiceMesh != nil && rr.DSCI.Spec.ServiceMesh.ManagementState != operatorv1.Managed {
-		// Delete servicemesh and serverless resources explicitly in
-		// this case, since the GC won't collect them because the Kserve
-		// CR generation hasn't changed.
-		for _, res := range rr.Resources {
-			if isForDependency("serverless")(&res) || isForDependency("servicemesh")(&res) {
-				err := rr.Client.Delete(ctx, &res, client.PropagationPolicy(metav1.DeletePropagationForeground))
-				if k8serr.IsNotFound(err) {
-					continue
-				} else if err != nil {
-					return odherrors.NewStopErrorW(err)
+	if rr.DSCI.Spec.ServiceMesh != nil {
+		// servicemesh is set to Removed
+		if rr.DSCI.Spec.ServiceMesh.ManagementState == operatorv1.Removed {
+			// Delete servicemesh and serverless resources explicitly in
+			// this case, since the GC won't collect them because the Kserve
+			// CR generation hasn't changed.
+			for _, res := range rr.Resources {
+				if isForDependency("serverless")(&res) || isForDependency("servicemesh")(&res) {
```

Comment:
@zdtsw see this condition: if ServiceMesh is set to removed, it's deleting _both_ servicemesh and serverless resources from the cluster. So if this is correct (it was based on my understanding of the previous logic), then the serverless resources should be removed from rr.Resources also in this "servicemesh removed" case, right? Otherwise they're going to constantly be deleted and removed in a tight loop.

Reply:
before this PR, if i understand it correctly: in`cleanUpTemplatedResources` if servicemesh is non-managed (including remove and unmanaged) resources with label dependency from servicemesh or serverless, should all be deleted + exclude from rr.Resources.
i think, for the removed case: yes; but for the unmanaged case: they should not be deleted but only removed from rr.Resources. => this is the biggest change in this PR in `cleanUpTemplatedResources()`

Reply:
Yes, agreed. Although I think this PR also stops removing the _serverless_ resources from rr.Resources in the servicemesh == Removed case (even though it deletes them from the cluster), right? I think that might need to be added back.

I can try to add a test to demonstrate the difference that I'm talking about (and add it back, if it makes sense).

Reply:
it should be covered in https://github.com/opendatahub-io/opendatahub-operator/blob/main/internal/controller/components/kserve/kserve_controller_actions.go#L288 if i am not mistaken

Reply:
That only removes the servicemesh ones, but there's a `rr.RemoveResources(isForDependency("serverless"))` missing there I think.

If the serverless resources _shouldn't_ be removed when dsci servicemesh is set to removed, then the `isForDependency("serverless")(&res)` should be removed from this line that we're commenting on. I'm not sure which is more correct, but the removal of both in this case is based on my understanding of what happened previously.

Reply:
It seems that DSCI ServiceMesh must be managed if Kserve serving is managed, so my point might be moot (old behaviour may have been redundant): https://github.com/opendatahub-io/opendatahub-operator/blob/main/internal/controller/components/kserve/kserve_controller_actions.go#L49-L58

Although, now wondering what the behaviour should be if Kserve serving is "unmanaged" and then DSCI serving is removed :thinking: